File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-7.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        count = 0
        curr = head
        while curr and curr.next:
            curr = curr.next.next
            count += 2
        count += 1 if curr else 0
        oneBefore = count - n
        dummy = ListNode()
        dummy.next = head
        curr = dummy
        for _ in range(oneBefore):
            curr = curr.next
        print(curr.val)
        temp = curr.next
        curr.next = curr.next.next if curr.next else None
        del temp
        return dummy.next

File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-7.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        unordered_map<Node*, Node*> random_map;
        Node* curr = head;
        while (curr) {
            random_map[curr] = new Node(curr->val);
            curr = curr->next;
        }
        for (Node* curr = head; curr; curr = curr->next) {
            random_map[curr]->next = random_map[curr->next];
            random_map[curr]->random = random_map[curr->random];
        }
        return random_map[head];
    }
};
